src/types/qwxdgshell.cpp

```diff
@@ -308,6 +344,9 @@ class QWXdgPopupPrivate : public QWXdgSurfacePrivate
         : QWXdgSurfacePrivate(handle->base, isOwner, qq)
     {
         sc.connect(&handle->events.reposition, this, &QWXdgPopupPrivate::on_reposition);
+#if WLR_VERSION_MINOR >= 18
+        sc.connect(&handle->events.destroy, q_func(), &QWXdgPopup::destroy);
```

Comment:
No, don't just emit destroy

> wlr_xdg_surface.events.destroy is now emitted when the xdg_surface is destroyed. To get notified when its role object (xdg_toplevel or xdg_popup) is destroyed, use wlr_xdg_{toplevel,popup}.events.destroy respectively.

@zccrs How to deal with https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4345

Reply:
xdgsurface reuse is rare, now invoke on_destroy of the base xdgsurface class when role object receive destroy signal (also destroys qwxdgsurface). in case re-used, qwxdgsurface will be re-newed